### ISSUE TITLE
DL4J: Loss function weights array should be automatically cast to appropriate datatype

### ISSUE THREAD

**Issue by Blacktoviche** [deeplearning4j/deeplearning4j#8431]
#### Issue Description
ClassNotFoundException: org.nd4j.linalg.api.ops.impl.transforms.floating.Histogram
Please describe our issue, along with:
- expected behavior
- encountered behavior

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version 1.0.0-beta5
* Platform information (OS, etc) Ubuntu
* CUDA version, if used
* NVIDIA driver version, if in use

#### Additional Information

Where applicable, please also provide:

* Full log or exception stack trace (ideally in a Gist: gist.github.com)
https://gist.github.com/Blacktoviche/14434b16cf02cfd8b027e8c2a4e2ffe5
* pom.xml file or similar (also in a Gist)
https://gist.github.com/Blacktoviche/1c99c1e7ac7276ef058d1b5d07585e65
#### Contributing

and If I fall back to 1.0.0-beta4 I got this :
java.lang.IllegalArgumentException: Op.X must have same data type as Op.Y: X.datatype=FLOAT, Y.datatype=DOUBLE

**Comment by Blacktoviche** [comment]
It seems library conflict with dl4j ui . when I removed deeplearning4j-ui from maven everything is fine
So I can't use deeplearning4j-ui 1.0.0-beta4 with deeplearning4j-core 1.0.0-beta5
So fall back to deeplearning4j-core 1.0.0-beta4 with deeplearning4j-ui 1.0.0-beta4 
and I solved the issue 
> java.lang.IllegalArgumentException: Op.X must have same data type as Op.Y: X.datatype=FLOAT, Y.datatype=DOUBLE

by changing weightsArray from 
 final INDArray weightsArray = Nd4j.create(new double[]{0.57, 0.75});
to 
 final INDArray weightsArray = Nd4j.create(new float[]{0.57f, 0.75f});

Now I can run my train/test model 
It seems we have to wait until the next dl4j-ui beta 5 so we can use it with dl4g-core beta 5

**Comment by AlexDBlack** [comment]
>So I can't use deeplearning4j-ui 1.0.0-beta4 with deeplearning4j-core 1.0.0-beta5

Right, that's correct. You can't and shouldn't mix versions. Everything - DL4J, ND4J, etc - needs to be the same version.
In fact, you probably got a big warning logged about that when you started your program - we have a version checking system.

You should always use the latest version too (for optimization, bug fixes and new features), you can see release notes here: https://deeplearning4j.org/release-notes.html

> It seems we have to wait until the next dl4j-ui beta 5

There's nothing to wait for. Everything is always released all at once.
We did drop scala 2.10 support in 1.0.0-beta5 (and added scala 2.12 support), so you need to use deeplearning4j-ui_2.11 or _2.12 instead. See the release notes I linked previously.

> by changing weightsArray from

Can you share code for this? It's a separate issue to the other one and I don't have any context here. Do you mean weight array in a loss function or something?

**Comment by Blacktoviche** [comment]
Thanks for your explanation. I should've looked in the maven repo for the latest dl4j-ui.

> Can you share code for this? It's a separate issue to the other one and I don't have any context here. Do you mean weight array in a loss function or something?

Yes it is the weight array in a loos function. here is  the code

>  final INDArray weightsArray = Nd4j.create(new float[]{0.57f, 0.75f});
final MultiLayerConfiguration configuration = new NeuralNetConfiguration.Builder()
                .weightInit(WeightInit.RELU_UNIFORM)
                .updater(new Adam(0.015D))
                .list()
                .layer(new DenseLayer.Builder().nIn(11).nOut(6).activation(Activation.RELU).dropOut(0.9).build())
                .layer(new DenseLayer.Builder().nIn(6).nOut(6).activation(Activation.RELU).dropOut(0.9).build())
                .layer(new DenseLayer.Builder().nIn(6).nOut(4).activation(Activation.RELU).dropOut(0.9).build())
                .layer(new OutputLayer.Builder(new LossMCXENT(weightsArray)).nIn(4).nOut(2).activation(Activation.SOFTMAX).build())
                .build();

I only had to change double to float to solve the issue which I mentioned before

**Comment by AlexDBlack** [comment]
> I only had to change double to float to solve the issue which I mentioned before

Ok, so it is a loss weights array. Yes, creating it as float (or casting to float) is a legitimate solution.
I think to make it easier for users we should handle that internally - i.e., both float and double should just work for the weights array.
